<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Sarah P. Flanagan" />

<meta name="date" content="2017-04-28" />

<title>An Introduction to Using gwscaR</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">An Introduction to Using gwscaR</h1>
<h4 class="author"><em>Sarah P. Flanagan</em></h4>
<h4 class="date"><em>2017-04-28</em></h4>



<p><code>gwscaR</code> is a collection of functions that are useful for population genomics analyses. They were primarily written while analyzing two datasets: one RAD-seq and morphological dataset from 12 populations of pipefish (Flanagan et al. 2016) and the other containing RAD-seq data from a single population of pipefish, which was analyzed using selection components analysis (Flanagan &amp; Jones 2017).</p>
<p>Selection components analysis compares allele frequencies among individuals in a population to identify genetic regions associated with different components of selection (e.g.Â comparing mated and nonmated males to estimate sexual selection). It is an approach originally outlined by Christiansen &amp; Frydenberg in 1973 (for use with allozyme and chromosome inversion data) and has recently been applied to next-generation sequencing data in the Mimulus guttatus plant (Monnahan et al. 2015) and pipefish (Flanagan &amp; Jones 2017). Monnahan et al. (2015) used a maximum likelihood appraoch to identify loci experiencing different forms of selection. In my study of pipefish, I compared the maximum likelihood approach to allele frequency comparisons using Fst values, an approach I had tested with a simulation model a couple years before (Flanagan &amp; Jones 2015). The comparison demonstrated that the two approaches find similar patterns.</p>
<p><code>gwscaR</code> does not implement the maximum likelihood method. This package only performs the Fst analysis, as well as some other useful population genomics analyses. Using <code>gwscaR</code>, you can:</p>
<ul>
<li>Run a Fst-based genome-wide selection components analysis</li>
<li>Plot genome-wide values (e.g., Fsts)</li>
<li>Manipulate vcf and plink files</li>
<li>Calculate pairwise Fst and Pst between two groups</li>
<li>Test for significant isolation by distance at each locus</li>
</ul>
<div id="genome-wide-selection-components-analysis" class="section level2">
<h2>Genome-wide selection components analysis</h2>
<p>The genome-wide selection components analysis uses a vcf file, so make sure your data are in that format. Then read the file into R.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(gwscaR)
vcf.file&lt;-<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example.vcf.txt&quot;</span>,<span class="dt">package =</span> <span class="st">&quot;gwscaR&quot;</span>)
vcf&lt;-<span class="kw">parse.vcf</span>(vcf.file)</code></pre></div>
<p>To run the program, you need to do a bit of file manipulation. First, itâs good to create a unique index for each locus and then create a vector with all of the column names that specify the locus information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vcf$SNP&lt;-<span class="kw">paste</span>(vcf$<span class="st">`</span><span class="dt">#CHROM</span><span class="st">`</span>,vcf$POS,<span class="dt">sep=</span><span class="st">&quot;.&quot;</span>)
<span class="kw">head</span>(vcf$SNP)</code></pre></div>
<pre><code>## [1] &quot;1.7022&quot;   &quot;1.42040&quot;  &quot;1.54720&quot;  &quot;1.118741&quot; &quot;1.245650&quot; &quot;1.985364&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">locus.info&lt;-<span class="kw">c</span>(<span class="st">&quot;#CHROM&quot;</span>,<span class="st">&quot;POS&quot;</span>,<span class="st">&quot;ID&quot;</span>,<span class="st">&quot;REF&quot;</span>,<span class="st">&quot;ALT&quot;</span>,<span class="st">&quot;QUAL&quot;</span>,<span class="st">&quot;FILTER&quot;</span>,<span class="st">&quot;INFO&quot;</span>,<span class="st">&quot;FORMAT&quot;</span>,<span class="st">&quot;SNP&quot;</span>)</code></pre></div>
<p>Also, you need to create vectors for the groups that you want to compare. In this case, compare males and females. The females have names with âFEMâ and males have names with PRM and NPM.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grp1&lt;-<span class="kw">grep</span>(<span class="st">&quot;FEM&quot;</span>,<span class="kw">colnames</span>(vcf),<span class="dt">value=</span>T)
grp2&lt;-<span class="kw">c</span>(<span class="kw">grep</span>(<span class="st">&quot;PRM&quot;</span>,<span class="kw">colnames</span>(vcf),<span class="dt">value=</span>T),<span class="kw">grep</span>(<span class="st">&quot;NPM&quot;</span>,<span class="kw">colnames</span>(vcf),<span class="dt">value=</span>T))</code></pre></div>
<p>Then you calculate Fsts between the two groups with fst.one.vcf and create a data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel&lt;-<span class="kw">do.call</span>(rbind,<span class="kw">apply</span>(vcf,<span class="dv">1</span>,fst.one.vcf,<span class="dt">group1=</span><span class="kw">c</span>(locus.info,grp1),<span class="dt">group2=</span><span class="kw">c</span>(locus.info,grp2),
    <span class="dt">cov.thresh=</span><span class="fl">0.5</span>,<span class="dt">maf=</span><span class="fl">0.05</span>))
sel&lt;-sel[!<span class="kw">is.na</span>(sel$Fst),] <span class="co">#Remove the ones that weren't polymorphic</span>
<span class="kw">head</span>(sel)</code></pre></div>
<pre><code>##         Chrom      Pos       Hs1        Hs2         Hs         Ht
## #CHROM1     1    42040 0.0950000 0.09613844 0.09556922 0.09556942
## #CHROM2     1    54720 0.3681519 0.39747241 0.38281217 0.38327248
## #CHROM3     1   118741 0.2648980 0.42850081 0.34669939 0.35852376
## #CHROM4     1   245650 0.2268431 0.15831111 0.19257711 0.19353493
## #CHROM5     1   985364 0.1829561 0.24357960 0.21326785 0.21407120
## #CHROM6     1  2253532 0.3802626 0.44444444 0.41235350 0.41539661
##                  Fst NumAlleles Num1 Num2
## #CHROM1 2.095737e-06          1   50  158
## #CHROM2 1.200995e-03          1   37  106
## #CHROM3 3.298073e-02          1   35  119
## #CHROM4 4.949101e-03          1   46  150
## #CHROM5 3.752718e-03          1   54  155
## #CHROM6 7.325786e-03          1   47  153</code></pre>
<p>To identify the SNPs that are outliers, compare the value of 2NFst(k-1) to the chi-squared distribution with k-1 degrees of freedom, and then adjust for multiple tests using the Benjamini-Hochberg (1995) false discovery rate.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel$Chi&lt;-<span class="dv">2</span>*((sel$Num1+sel$Num2)/<span class="dv">2</span>)*sel$Fst
sel$Chi.p&lt;-<span class="dv">1</span>-<span class="kw">pchisq</span>(sel$Chi,<span class="dv">1</span>)
sel$Chi.p.adj&lt;-<span class="kw">p.adjust</span>(sel$Chi.p,<span class="dt">method=</span><span class="st">&quot;BH&quot;</span>)</code></pre></div>
<div id="the-wrapper-function" class="section level3">
<h3>The wrapper function</h3>
<p>For ease of implementation, all of the above steps are contained in the wrapper function <code>gwsca</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel&lt;-<span class="kw">gwsca</span>(vcf,locus.info,grp1,grp2,<span class="dt">prop.ind.thresh=</span><span class="fl">0.5</span>,<span class="dt">maf.cutoff=</span><span class="fl">0.05</span>)</code></pre></div>
</div>
<div id="plot-the-results" class="section level3">
<h3>Plot the results</h3>
<p>The <code>gwscaR</code> package includes a function to plot genome-wide statistics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lgs&lt;-<span class="kw">seq</span>(<span class="dv">1</span>,<span class="dv">22</span>)
<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))
sel.plot&lt;-<span class="kw">fst.plot</span>(sel, <span class="dt">fst.name=</span><span class="st">&quot;Fst&quot;</span>,<span class="dt">axis.size=</span><span class="fl">0.6</span>,
                   <span class="dt">xlabels=</span>lgs,<span class="dt">xlab.indices =</span> <span class="kw">seq</span>(<span class="dv">1</span>,<span class="dv">22</span>),
                   <span class="dt">chrom.name=</span><span class="st">&quot;Chrom&quot;</span>,<span class="dt">bp.name=</span><span class="st">&quot;Pos&quot;</span>)

<span class="co">#plot without the scaffolds</span>
sel$Chrom &lt;-<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(sel$Chrom)) <span class="co">#the lgs and chrom have to be the same class</span>
sel.plot&lt;-<span class="kw">fst.plot</span>(sel, <span class="dt">fst.name=</span><span class="st">&quot;Fst&quot;</span>,<span class="dt">axis.size =</span> <span class="fl">0.6</span>,
             <span class="dt">chrom.name=</span><span class="st">&quot;Chrom&quot;</span>,<span class="dt">bp.name=</span><span class="st">&quot;Pos&quot;</span>,<span class="dt">xlabels=</span><span class="ot">TRUE</span>,
             <span class="dt">scaffs.to.plot=</span>lgs)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAACBFBMVEUAAAAAAAoAAAsAAA0AABQAABcAABkAAB0AACAAACEAACIAACgAACoAADEAADIAADQAADoAAD4AAEkAAFgAAGYACgAADQAADSAADVEAECgAFwAAF2YAJAAAKJAAKhAAKjoAKpAANbYAOigAOpAAPioASUkAWLYAZoEAZrYNAAANWNsRIAsTFAAUAAAXAAAXAA0XfNsYIAAdAAAdOgAdkNsgAAAgABcgIQAganwge9shIAAhfNshhv8kkNsoAAAoAFgogf8okNsqAAAqADoqkJwqkNsxABcxkNsyFwAyUTEykNs0AAA6AAA6ABc6ADo6AFg6AGY6Zj46Zlg6e3w6kLw6kNs6nf8+AABJAABJgWZJtv9RvP9YAABYOgBYWCBYZjpcKABmAABmABdmAGZmOgBmOpBmPgBmXChmZmZmgWZmnZBmtv9m27Zm2/9wkDp7FAB72/98IQB8OgCBKACBZgCBtmaB22aB25CB//+QKgCQOgCQOmaQhTqQkLaQnGaQnWaQvJCQ27aQ2/+cOgCcZheckDGckDKckDqctmac//+dOgCdZhedkDqpqam2SQC2ZgC2cBe2nTq2tli2tma2vFG225C2/9u2//+8///T09PbSQrbZhfbkDLbkDrbtmbb////gSj/nDr/nTr/tkn/tmb/22b/23v/25D//4H//7b//9v////1PVx8AAAACXBIWXMAAA7DAAAOwwHHb6hkAAALIElEQVR4nO2di5/cVBmGj5ZRcRCHbkQEFXQFV9R6RcVLlSJel4orq7X1wmq9rENVpN5WLboqFSWMutIh1nFbTB1380/65X5yfTObzEx2+j4/l2aSk0zOM9/3nXP2EpVDSlHzvoG2Q0EACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBCgRNFC9bX9r79tDZ39LqeXd2dxUmygWZC/v2r4RWxlDx14RZf1Z3VZ7KBY0Wvcix3H2f7P/q6G/q589Xy14kpYI6juhmOBfN4gmucJCUCGCQkGDXD/Xr6C4BnmC9rf6k15hIUCj2N53dn1BA6k2an2yKywC9btHQVO/QquhIAAFASgIQEGAeQuyrNo3MF3mLMiy2m6IggBMMcC8BbUeCgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgwKwEmWbtN5oPMxJkmofVEAUBZppirf+1+hxmWaTb/4cZObRFUGvVzXSYL/PTVkMtmQdREKKtflojqLVQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhQ0r3occn+xkgpPstVJ3pUqb9R+Ljk61ZQ9LBbf2P0/dwI4uOSww33WcA5z5NmBEUR1Jdk48NuNXJqECMoQfS45GgUywug61jQzK7QaigIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBGhA0GH9K4xqNCJokQ1REIApBmCRBkxfUGv/yKAaUxdU+89U5iy49YLm/XdArUqxvHK/+IKqk/+Xv4ueYklKe9vKP42erSCQLy300y5BbaRNKdZKagsy4ytUzpBYU/uF1RVkmsqMNisaihPtEKQcBQGmmmJFwq6nFCu7QivnNZMyA0Htj5IyGhKULyHwMxtDU3qXZgSVP9dlJoKm9TbtFjRBDWu3oNL4Th6bpG5PVOVbnWJlJD/aoj7ndq8Nw2ATP9UoP15JUEGCzN9PExNF0Aut7+5GUQB5jZoR0miyTVVQapiHzzBrJqWaLddTSzG5yfREMbrzog4sqKB4U+uee5fp/np3LruKe1DPTxypda6SoklBupBQRoIgqkJBDZfg6cyEpiVIK8jZUSz0EzRvSFShoFremhKUGYSiGmTpVVpvEwRTc7OdQj91DDU2k07dRFyDLK02pyfVVl6papo2CLJyBaVTLHunVm6papoWpFjOPC/vrrL7ksrQ+fOgycVqYa5U7GxmFJzs9OnQ5G+YFS9Ey7sYnlUgqO7wXS+DG/wdxRw/URGq9iP5/BSrKajmGNCQoGC0Tt1JuCc5EUo0SQzz4cokvePQC9IXFnq11u4tL420USyuYcGVTE1dgZ+q/Z57isX9DNdZ0XQ5W55Ss+0gOqKzYkFwhjSj76Y1sppPCYo6GDfKrTNmNH2Kz08u2PyJZvL9wteHRZAZpliihIRz5LhZtjNmLCgIt6DLUYr5R5KG4tez+XZjc4LCz9vb0NcQYee9zcRSLFuCvMPZNNTebzpr9mKa+OUFv6NWFDhO0K14LRatW/VUiwtNGIBxCGnvkA2hOvc78dnNCXJ0QdpaLE+QvtSPz8paPFifSpg8/hpLMS0mHK2E6IKiqEokpKmFkBMqq1peqjXMX8FUowlB+jCtjWLaPM/KCorjLRyqokOxIdT/jMryH6/pn1tVmvkFqqygZA2Kss3JCPKukjNNiHpW9pmnBWUCJLGIPtDEoOEaVCAoOQWIAyZVsUx9Ju3ETcoMJV4WzAmKF4uYpgXp86Cox3F0pHblC0qkmC4I50fet+30m528g9OIIEdPFTOxFUuIRy+ttCciSDuU290qFJ9Q9VJTGea1iWJKUHKdZekbTtaiiQRVCIkCEZVlNyLIsdJdjgU5yc2MKS3VCiUWdbZCUSkSEe1HV5jGTDpbg5LBURBB+houeajoB9a+oNIuBmdn20R+gKGGBZnp/qWSLd4VrcWsxPl6ZsWTJiv+zog/w4rePW+oz4rQ2yRbm4nr5VBdUPT05DqC4s7rrbLnpy9ghYJ8zVpchFpjIe5hb1fQ9fDN9W8VRIOAFR2oKyh6cmmpoIL+laRYwlj2UMaPGfiJqnoYZZEgM5CeWANHc4fopCh2y7OssqDo2bep81UVQVZ8SwcWFHwe/oeeEJT6IVG5oOisxgWFT09OoX2C6Q6mj8UL0VxDTngocSF/qRH134+CRIp5VhLzwZwU078VlUgxK6+CH0hQfgQtPHVr0MJTbxQbP3JpS6rQ+kj5BwcqwTfl6wefUqpzzrHdFuOTHz8rux7ymo8fOR+2e0t8ymfk6x3ydb9c/Atu49xHNM+QWvOgvY2lsyJmpI4OHVs2bPWKH7uhJgdW5OvNsus/qvP4Ru/E5k2rxs5NqyLmdxsd9epT7ouutN7b6D3j/O/40eHehlLd7X8p9dH9H6k3OZfVPbLHu3bu/xXD7KgjyI2K433x9LWlobO/teJ26Zn9LfczHxnD0dLZFbfJy5Qynn6N/GfHjZCvKnX0K/Li7+6L3p+UWhpKm+5P3ZcrzrV3qpX/3vd2kfbnl8uO0/6VmunpAakhyFbrtn/3bvSIoNHyT+7YHZ/0E3G8+ujWiq3WlHrANv6q7rW7Sr2y25VzuurmC131wTf2vv469aWHxZu7a/2qeo/47Kh7xqvfO3LG+ad6t39xO296OkPqLTUCQQP5Z9T77cntJ18rn/snN9zKIftG0rcryhjaxt/Ul6Wp3e1KQ9s4b+wYF2/8gPGY2/qWU0qtye7nuj/sbUswnR6f/K6EznPdX/gXHxzeCHJCQW4oKfXo1vreqdtekAL8kITTWzfWvRy88T5X0B98QR1PkITLFaU+vHb0817r9zv7v3RrkJdr6l43ilTfVjdvexe3512lGxBke3V0/4nbXnAGy7uj3lOeBYke98Mfn7jhnN07oQm6Is1t44+33mmcl9b+LgmXjhtY6vXqIyJPbaq7e54ge941ugFBwWc8fthVIl02nl/ti5c37I7lX0fGI6nBP3vXmTDFbCnLUlgefKrjCer2TvSdq+r04E4Z6uRKv1fqhse/5VU393/zjp8GBI2CPoyCicym8byU6WvH7g/K65XOOdv4x9tOu01F0EWv4Lhp5Lfu9h7cdq7KVGCte8S90mX1odVvqAfcwcs2np2/n9qCnlUSLl6oiIHx8bt29zY2xcCLUmf93W76GE/f+mk5PLi725HyIjHjhVPvvW5rP8V+vrp5RH3C1WxcekLdtesMvCGtX7+DdakpqHfc/Ywlsxxv/vPkHbvO6PbPSdl91bY/AF07dovr4kLvgrHTu9CRAi6j2PK/jUtba8Zj0vqy+pi0Xtrp/Vq9zxheO6bWx6svPeNF34sSSPOnpqAwVYS+vB7c7lVYpT7rD9Gy9UXlLTW81YX9Eq25iHJby/RIsu6i8ZfkKkVaXw43mujngVnwh/zVh4IAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQgIIAFASgIAAFASgIQEEACgJQEICCABQEoCAABQEoCEBBAAoCUBCAggAUBKAgAAUBKAhAQQAKAlAQ4P/snTB8WJ7UdgAAAABJRU5ErkJggg==" /><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAASFBMVEUAAAAAADoAAGYAOpAAZrY6kNtmAGZmOpBmZmZmtv+QOgCQOmaQkLaQ2/+pqam2ZgC2///T09PbkDrb////tmb//7b//9v///+TgcVIAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAHY0lEQVR4nO2c4WKbNhRG2bpua7eR0qbR+7/p3NiAdHXFASNj2fnOnzqKENJB90rQmC6IRbp7d6B1JAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAiQIECCAAkCJAhYEPTS/fbf+dPPP7+Ft69d9+nHMZ1qibKg108/Xs9GXrvfv4XXP07KvhzWr2YoC/r++X3mhPD279s/385FuaCue/IgXRD0JYxiLv/+mkRbWngKVsygUdCL6+fjCppz0Lugt6+FBPRhBZ1XsZ9//TgLejllm+7zthaegf3Dk6Cbt9A0EgRIECBBgAQBDy1oGG5/jkcWNAwHGJIg4JEFKcRaQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIKAtgT1fcXG6tCUoL5vz5AEAU0JmkPsiL+hX0dbgkYO+RbGOp5J0E2ktinoqrHeZto1KugaJIj4SCHWDhIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBCwMLzpdcnnD9+7Tu9yjZleVXr+UHxd8ocVNL3s9vzh+9/uDNLrkscPv94F7LxPWjNomkEnOa962W2Ek4M0gxKm1yVPq5g3gT6woMNaaBoJAiQIkCBAggAJAiQIkCBAggAJAiQIkCBAggAJAiQIqCGovS9YVKSCoBa/glIPCQIUYoCSNNCAoGa++eRyf0H3/W4YnvuDC+KT31/QTUJs7brxEIJuwPqdxwOEWJEdM6vi1qxdQbtyU72t2ZMKqke7ghrZH+0eXu+2UP3uI7d1kL+9gvq+c1xUv3/N4+2oCJQg4F4htlXgw4bYlS08zjOkFgS1sVwVqCloy0ATPy0bqijoyoEOErTMffysToH3CrGIRUM3yuXrF4kWbjUWDG1a7TZcoEMF7b7IlQRtitUjQ2zvlsYd2Fi0aQJF7bT0POg6QfNBnqCrMnfip5aho0NsnhmxIadafu+1pVNNCXJLC/2bBg4jiOqZgpW0FGJeYWn88zhhBNZjKujAG7mDBc2Rkt5rlBqf20n9pK3f0lddQfMoSn12Qsy/1zAhlvB+8KnCunDdR1VBnCgmGSRoual3P15b9al7q7FeEIUYNhUJepgQW5NIVy9GWG8OsZyise27q8ohNn/eOO+r3tEvLBHeeRbPfZu/MDuVbBO085mHOdU2QcvnvsnfKL6XgJ/01/sEZT3YFGLHCYr2OCv+V8MaKlRbbKW4zptXm+fXa/Uj34ohFu+SvRutpIsFh2tjJVySdLKMJRco7lPf9+Yqrv9Pg8qPXDND8V7FdNFpK98iLwRLtM4np0rPZQRdKt3lgdmSoLyLLiuTya/SzI8RNBVOJcnZvTsYf3jwe6azMuLzBzOB5rgozSA+YX+508hMxxmpzwLaJsqpy8uKagsyW9ssSwyTHydNmcwwuBr7TJCXgdK5aH6ccTTb4S39cg19GmJpCvSiIBNUSu4nnHFNYzdnS6avI6jQ/WMEWRtJxBemtPPItV8U5D8GmI4aEkOzkrFKcQDHhtjYwczQuBMp7GTnRJUW5sthsBPWTqgQT5hz6SUrXTe8q46KmEPMxEqyOZomfVQlRIe5z0CCF6xRQ+mCZEIsaTqLufVPAGr+AZUjKB1WlFrtZHGegYRgBAXbUtKT0uUxu4ww1xmrLA6whiB/H+QIirJ0Hk2XalNhOg67SOX7KU9QOoHTfheypKX2KhaNYexA3J9xuvRzWhjMcdnA4oKoDAQlhnJBwcyxIvVmkB2oXYujSkO8V0lT0JKgpCxLXPbsUUOpn+mwNX7qCIo2gWOpk0mnFOQJMoHgl0RHBntQdnkiAVkzieblAdYT5E2FJOWcsD9HdXIfSYFV69zERJcnXfwHc8FswQGCTDLJB59c9+wBRKHQxmoYBjs557mQG4ovWnz2JCnF8/4mgsJFUHp/bS98NoPK2SXP5V4EBTPY7OyxIHP/5k3NGoKmtyenfiZBw+Cc30wOUyENQ6sjnnn52EyVPFLjHJjV6fOj9gqa3lzqCwqxoN6c3/YnTqTOwObCTOM8qUyV6Wdvhczq2IKFca8WNL371hzf3VCQ3S9sExRQUByX+wWNb082TJ3JclCcOe3P2dCDUxjyVBW1nbqO0rjdpWfH+Jl9tyB/Bj09e3PQ07N3FXt6WvxSd1NIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQB/wP/k7O1MRqjtwAAAABJRU5ErkJggg==" /></p>
<p>If youâre comparing different sets of Fsts from the same species, you might have different loci represented in each comparison. But you might want the graphs to have the chromosomes line up nicelyâ¦so you can take advantage of <code>fst.plot</code>âs <code>group.boundaries</code> option</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#create a data.frame with the starts and stops for the chromosomes</span>
bounds&lt;-<span class="kw">data.frame</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(vcf$<span class="st">`</span><span class="dt">#CHROM</span><span class="st">`</span>)),<span class="kw">tapply</span>(<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(vcf$POS)),vcf$<span class="st">`</span><span class="dt">#CHROM</span><span class="st">`</span>,max))

<span class="kw">par</span>(<span class="dt">mar=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>))
sel.plot&lt;-<span class="kw">fst.plot</span>(sel, <span class="dt">fst.name=</span><span class="st">&quot;Fst&quot;</span>,<span class="dt">axis.size=</span><span class="fl">0.6</span>,
             <span class="dt">chrom.name=</span><span class="st">&quot;Chrom&quot;</span>,<span class="dt">bp.name=</span><span class="st">&quot;Pos&quot;</span>,<span class="dt">xlabels=</span>lgs,
             <span class="dt">scaffold.widths =</span> bounds)

<span class="co">#Or if you just want to plot the linkage groups</span>
lgs&lt;-<span class="kw">seq</span>(<span class="dv">1</span>,<span class="dv">22</span>)
sel.plot&lt;-<span class="kw">fst.plot</span>(sel, <span class="dt">fst.name=</span><span class="st">&quot;Fst&quot;</span>,<span class="dt">axis.size =</span> <span class="fl">0.6</span>,
             <span class="dt">chrom.name=</span><span class="st">&quot;Chrom&quot;</span>,<span class="dt">bp.name=</span><span class="st">&quot;Pos&quot;</span>,<span class="dt">xlabels=</span>T,
             <span class="dt">scaffold.widths =</span> bounds, <span class="dt">scaffs.to.plot=</span>lgs)

<span class="co">#And you can highlight the points that have adjusted p-values &lt; 0.05</span>
<span class="kw">points</span>(sel.plot$plot.pos[sel.plot$Chi.p.adj&lt;<span class="fl">0.05</span>],sel.plot$Fst[sel.plot$Chi.p.adj&lt;<span class="fl">0.05</span>],
       <span class="dt">col=</span><span class="st">&quot;cornflowerblue&quot;</span>,<span class="dt">pch=</span><span class="dv">19</span>,<span class="dt">cex=</span><span class="fl">0.75</span>) <span class="co">#only one point is significant in this example</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAASFBMVEUAAAAAADoAAGYAOpAAZrY6kNtmAGZmOpBmZmZmtv+QOgCQOmaQkLaQ2/+pqam2ZgC2///T09PbkDrb////tmb//7b//9v///+TgcVIAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAHRElEQVR4nO2c4WKcNhAGadM0TdrikDjm/d+09gHSSkgargiQz9/8SAjHydqJdiVxPrpRFOmu7kDrSBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBBQEPTU/fbvdPTrz+/jy7eu+/TznE61RF7Q86efz5OR5+737+PzH6/Kvp7Wr2bIC/rx+TZyxvHln5e/v0+n1oK67sGTtCDo67iImf9+G0T3tPAQbBhBi6CnpJ+PK8jXoJugl2+ZAvRhBU2z2K8vPydBT6/Vpvt8XwuPwP7wJOjwFppGggAJAiQIkCDgakHDsLsDx3KxoGFo3ZAEAUox4GpBzSNBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBwFmC+n73D7qGkwT1/Xs1JEHAqSnW/K/VJzizSLf/xYwEjQhq19yp03zBT7OG2lgHSRDRrJ9WBLWLBAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBBTCc49Lng5+dJ2e5WpxjyqdDrKPS/6wgtzDbqeDH38lR5Ael7wcvD0LOPE8aY0gN4Je5TzrYbeGRA3SCApwj0t2s1hqAH1gQae10DQSBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQUEHQe/0WxjaqCHpkQxIEKMUAFWngeEHtfstgE4cL2v09lYsFNy/o6i8CNZViqXL/+IK2k/7m76OnWEA52Ca/Gn3yVzLJ0N7O1KcpQS3SUoo1yW5BvW9hc4Y4T+9A2F5Bfd/17nCjIZdp7yHlJAg4NMVywj5SipVaaHJdcy8nCHoHw6RAJUFpCbOfcwwd9FPqCCpJOEnQUT+mbUF31LC2BRXHd/jaPXX7rirfdIqVCP9rMzHnatj102CNTzXKr28RlMuP6/3UWChCFCb4t4PMAJquqSOkarIdKiia5vEZZnVSqm65PizFXjsZLxTNHizd1IMK8ocmvLdexvHeev56Kh/BPj9+pO5pJaKmICtkkREwj6q6Fce0fsRMf5QgU5CDbt+uiSpOJVF5QXvE1RJ064ON1NWg7Cy2jKVqq528nx2Gqq2koz74GpSfxW5+1qWqOg0IGpKC4hRLXpQoVdW5PsXm0KMUW7E6F6+yixdfQ83NajZXtsUavj0qXZdR8zfMcoI2fqCaE7R3+t6XwRV/R3HtZ/43RJiZ7vOl6z52zgGVBM2zddgTdyJfaMZgml92JvGJdy/IbixMtbZd8yHHeTTvPka3iBz6ZX1Qvuu/Ne7LU8zHueyzejeofCdTeTgsywP3LiOIVkgn3U2rspuPBPV2wzURpdF86JZPdkUdtpRcG0RNHkqd+0FToDa+YYlvuWwdTO8FuRpmh5lxZ3EnzrndWE/QlFHzgd1DTO7mw9VWzJWg3huKbwsEho7ZtGep8csLZk9li4rdi0Vld5x9zFctA9AI8j8hMYR29PfuN9cTNFpBZi/mqlI4Lzl7/l1ri/8vpgL3D79qKWbGxNi7ez6jFeRGVZCQVtC4KKs7hUf38e6ihiA7TZtZzKzzhrUgP96sYH/1pntpK5PJ8G1rl6SYrTM2dl+DXLatL7q1klgm+GpdiCkWlK5Xqdq/mco1KCMoXAL4iS2qWL4qmfm+mBWrAZSc8bKbxQ1UFtTbau0rT7SLMHkYCQp2LeMYfQbC+ZFbM7nW7uaQWSxIlUBVKGEpODax7Jy3LCbT4W4h/4atTR0kaMgIWhdpK2o+Dq4uCtowJDIiNsuuIsjPPYMJywQaDiF/xp7ws+BaXibYDUUlJ8KdpxaOWEmnx40fHOuLpu6mDbtQ1pHOgkoxzu9eX7J1y3uQoDC+qAalry6MsrdglpEUzPuzn9LCe3W3PBTSTyu2PNsFuacnlwTZ3VUsaF1yEoJW18wvDoMbZe74FuvStBcy9Lc98dsfzk7fL/vk6b/BbZ0H98JeQe7JpTlBpfggxVCi9ePOLPt/m4a3qGfp5oPb6Z9uvC1j1S7Xdgtyz76N3t+dJmgc7XbECApLMQjy76otaHl6coTr8rgKMH7N3UXNGhpvsY62gWmf5gT084e1JsWmE7ZLpRSb3uZTbICJbO8Ienj21qCHZ+8s9vA8+JM39iNBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgRIECBBgAQBEgT8B5H2tZjXuVmkAAAAAElFTkSuQmCC" /><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAAS1BMVEUAAAAAADoAAGYAOpAAZrY6kNtkle1mAGZmOpBmZmZmtv+QOgCQOmaQkLaQ2/+pqam2ZgC2///T09PbkDrb////tmb//7b//9v///8fqghUAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAHXUlEQVR4nO2c7XqjNhBGabebtrttSZzNhvu/0mbjIDTSiANGYNl5z5+1CdbHYWYk/KzpBjFLd+0BtI4EARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEARIESBAgQYAEATOCHrvf/ju/+vnn0/D6veu+/DhmUC1RFvTy5cfL2chL9/vT8PLHm7Jvh42rGcqCnr++R84wvP77+s/T+VAuqOvuPElnBH0bRjEf//4KojUt3AULImgU9Oj6+byCphr0Luj1e6EAfVpB51Xs518/zoIe36pN93VdC/fA9ulJ0O4tNI0EARIESBAgQcAtC3p4eNi/kxsW9PBwhCEJAiQIuGFBqkFNIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQIEGABAESBEgQ0JSgvq/XVi1aEtT3DRqSIKAlQVGKnU71Wt1GU4ICp1Mzhu5I0D5O2xR0yWR3irpGBV2ABBGfKsXaQYIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQJmphcel3x+8dx1epZrTHhU6flF8XHJn1ZQeNjt+cXz324E6XHJ44tfzwJ2nietCAoR9CbnRQ+7jXBqkCLIEB6XHFYxL4A+saDDWmgaCQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCKghqMFfWNSjgqAmf4NSDQkClGKAijRwfUHN/PDJ5+qCrvzTMOz8kwvi3q8uaJcUW7xs3IKgHVix8Wg/xYpsiKyaO7NmBW2qTRV3ZvcpqCLNCmplf7R5er3bQvW7j1zXQQK3Cur7znFR/f41T7ijUlCCgCul2Gp/N5til7VwQ18htSCokfXKp6agFRO1flo2VFHQhRM9SdA8V/KztAheKcViZgXtVcwXLxMN3GrMhdCa5W7V9TlS0NaLXEnQylQ9MMW27mncmY2H1gRQ1ExT3wddJih+4p23TV5f0IyfaoaOTrEpMuaeCZgv/et8tSXIPVoYYJj4/BSi05IDC2kqxbyDpflPE52fQqrRCjryRu5gQVOqRH8vR8fUjPVjG9/VV11B0zRKg3ZqkHuvkaSY4fzZhdm6laqCuFJ4s/IEzbc0+jmfczuCFtx2ujXoomV+EnQ7KbZkq7Z0MeLzZhyWja3fXVVOsfByZdxXvaEv9+1Lne18n/9h1vfrBG38yiPpaqWg+c53+T+KS/yYv28TlHW2LsWOExTtcTB8klNKQ1y3ncw/NpXxssPdBQ3Bz/yNlhmi79ALxUK3b+0my5i5PmZIfd8nV9E0vHsNCv04huLNSjLEvCknV+aqib30ZotVFBQdWTm9i+liFyVB+RA9ZnIlPZj5SQSFg9Mh07t3C+NPD/7OdKmMuP8hCaAwr9I9xIIO31vK/NiKlN2t5YUyDHleUW1Byd42KxPBj1MDkoO+xo+5ZhtpW4GSWCyuqo7ndHpzf1xCb1PMlkAvDbIa4N2fnc90phXmnnRmojffZszU+iMEpTZMxmch/XFKPkRXkFdN4rbi3B2HFKUY388em2LjCDND406k8PXqVKiig86Cbb9jjJLVCkq/KMhDavn0LvpUxJRiixZat4xnqTlkR4xEU4OmE0LLJqbTXUbaP1DzP1A5gorzSoLFSbwhOytqqHfWMBO/U9Nx9/G4M7E+NQQ5i5gvaLKRZlP43DAetNOwS92oOh6IJ8jbpg7TSXbQBWqvYtEk0hGGcz4EDWNVymq7TUwzVaM6HkciyNun2nH762xGvQg6pYJ6WwTMSWOm9FkeFOZlcrNP3kcfSVb5uLdxyKG3JX7qCIqCYzzqCAolyBHkx0sm6JQVs0yz2SgmJw3JFVryPUs9Qd6akQpx/ERFeubKn1JBqUNzdezin2qcmjkugsxockHJhberSipxOhhly/tbZ6pT4coMmWsWh51RH8f9LoKGD0H2BrskyAmWNIBia1OgJRk0JHPNd1N2jYpbcfuvISg8Pdn6CYJOJ6f/JDqSt16GmYNWUHztE/XRCUncOYKiaI0+tlVQeHKpL2iIBaXzSseT1+h4YtPB7MB0qPTeadvuLs1J7Ge5oPDs2+TzXTVB+aXPQmqNoIEFmV38ZkHj05MTwmjyqQ5JEUjfZ3tZezD6WFbK/QVqSO5iiht59x53qyA/gu6erTXo7tm6it09Df6ouy0kCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIIACQIkCJAgQIKA/wESjfWK49DrtAAAAABJRU5ErkJggg==" /></p>
<p>Notice that in each of the examples above, the xlabels have been specified in different ways. The <code>fst.plot</code> function has a number of ways to specify plotting parameters so please see the documentation for more details.</p>
</div>
</div>
<div id="other-useful-functions" class="section level2">
<h2>Other Useful Functions</h2>
<p>The <code>gwscaR</code> package can be used in a broader context than just genome-wide selection components analysis. It also contains useful functions for performing typical population genetics analyses and file manipulations. Here I will highlight some of those functions.</p>
<div id="merging-vcf-files" class="section level3">
<h3>Merging vcf files</h3>
<p>In the above examples youâve already seen that <code>gwscaR</code> allows you to easily read a vcf file into a data.frame format. The package also allows you to merge two vcf files using <code>combine.vcfs</code> in case you have two groups of individuals that were genotyped at the same loci.</p>
</div>
<div id="choosing-a-single-snp-per-rad-locus" class="section level3">
<h3>Choosing a single SNP per RAD locus</h3>
<p>Many population genomics tests have the assumption of linkage equilibrium among loci, an assumption which is violated by using multiple SNPs per RAD locus. Therefore, the function <code>choose.one.snp</code> allows you to randomly select one SNP per RAD locus from a vcf.</p>
</div>
<div id="comparing-each-snp-to-a-distance-matrix" class="section level3">
<h3>Comparing each SNP to a distance matrix</h3>
<p>The <code>fst.ibd.byloc</code> function is helpful because it allows you to identify loci that are significantly affected by isolation by distance and distinguish them from loci that are not significantly isolated by distance. This function simply performs a Mantel test for each SNP. It uses data in ped file format and you must give it a list of the populations youâre comparing (this ensures that youâre comparing the same pairwise relationships represented in your distance matrices).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ibd.by.loc&lt;-<span class="kw">fst.ibd.byloc</span>(sub.ped,dist,pop.list) </code></pre></div>
</div>
<div id="plotting-structure-output" class="section level3">
<h3>Plotting STRUCTURE output</h3>
<p>The popular population structure program, STRUCTURE (Pritchard et al 2000), assigns proportions of each individualâs genome to population clusters. A companion program allows you to plot the graphs, but that program didnât allow me the customization options that I desired (including embedding a structure plot in a multi-figure graph; see Flanagan et al 2016 for an example), so I wrote my own plotting function, <code>plotting.structure</code></p>
</div>
<div id="calculating-pst-and-comparing-it-to-fst" class="section level3">
<h3>Calculating Pst and comparing it to Fst</h3>
<p>Sometimes you will have both genetic data and phenotypic data, and you might be interested in comparing population differentiation at both levels. One way to do that is to calculate Pst, which is analagous to Fst but with regards to phenotypes. <code>gwscaR</code> contains a funcion to calculate this metric for you, <code>pairwise.pst</code>. Once youâve calculated Pst, you can compare it to Fsts using the funtion <code>fst.pst.byloc</code> and to a distance matrix with <code>pst.mantel</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fem.psts&lt;-<span class="kw">apply</span>(fem.phenotype.data[,<span class="dv">3</span>:<span class="dv">10</span>],<span class="dv">2</span>,function(x){
    pst&lt;-<span class="kw">pairwise.pst</span>(<span class="kw">data.frame</span>(fem.phenotype.data[,<span class="dv">1</span>],x),pop.list)
    <span class="kw">return</span>(pst)
})

fem.pst.fst.loc&lt;-<span class="kw">fst.pst.byloc</span>(sub.ped,fem.phenotype.data,pop.list,<span class="dv">1</span>)

fem.dist&lt;-<span class="kw">.pst.mantel</span>(fem.phenotype.data,dist,<span class="dv">1</span>)</code></pre></div>
<p>For more detail on this method and to see an example, please see Flanagan et al. (2016).</p>
</div>
</div>
<div id="citations" class="section level2">
<h2>Citations</h2>
<p>Christiansen FB and Frydenberg O. 1973. Selection component analysis of natural polymorphisms using population samples including mother-offspring combinations. Theoretical Population Biology 4: 425-44.</p>
<p>Flanagan SP and Jones AG. 2015. Identifying signatures of sexual selection using genomewide selection components analysis. Ecology and Evolution 5: 2722-2744.</p>
<p>Flanagan SP and Jones AG. 2017. Genome-wide selection components analysis in a fish with male pregnancy. Evolution 71: 1096 - 1105.</p>
<p>Flanagan SP, Rose E, and Jones AG. 2016. Population genomics reveals multiple drivers of population differentiation in a sex-role-reversed pipefish. Molecular Ecology 25: 5043-5072.</p>
<p>Monnahan PJ, Colicchio J, and Kelly JK. 2015. A genomic selection component analysis characterizes migration-selection balance. Evolution 69: 1713-1727.</p>
<p>Pritchard JK, Stephens M, and Donnelly P. 2000. Inference of population struccture using multilocus genotype data. Genetics 155: 945-959.</p>
<hr />
<p><em>If you run into any problems, find any bugs, or have other comments on gwscaR please contact me: <a href="mailto:spflanagan.phd@gmail.com">spflanagan.phd@gmail.com</a>.</em></p>
<hr />
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
